### PR TITLE
Add EIP-5982: Role-based Access Control

### PR DESCRIPTION
**Pull request by xinbenlv**
When opening a pull request to submit a new EIP, please use the suggested template: https://github.com/ethereum/EIPs/blob/master/eip-template.md

We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:

 - The PR edits only existing draft PRs.
 - The build passes.
 - Your GitHub username or email address is listed in the 'author' header of all affected PRs, inside <triangular brackets>.
 - If matching on email address, the email address is the one publicly listed on your GitHub profile.

### CODE REVIEW

**Review by Pandapip1 — CHANGES_REQUESTED** [review]

File: EIPS/eip-5982.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+status: Draft
+type: Standards Track
+category: ERC
+created: 2022-11-15
+requires: 165, 5750
+---
+
+## Abstract
+
+An interface standardizing role-based access control.
```
[the remark this answers is not in the document]
Comment (xinbenlv) [reply EIPS/eip-5982.md]: Updated

File: EIPS/eip-5982.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
+---
+eip: 5982
+title: Role-based Access Control Interface
+description: Interface for Access Control
```
[the remark this answers is not in the document]
Comment (xinbenlv) [reply EIPS/eip-5982.md]: Updated

File: EIPS/eip-5982.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+## Motivation
+
+There are multiple ways to establish access control for actions requiring different privileges and powers. This EIP standardizes such
+interface so smart contracts, accounts and DAOs could establish access control based on given interface.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
```
[the remark this answers is not in the document]
Comment (xinbenlv) [reply EIPS/eip-5982.md:26]: Good point

File: EIPS/eip-5982.md [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+Interfaces of reference is described as followed:
+
```
[the remark this answers is not in the document]
Comment (xinbenlv) [reply EIPS/eip-5982.md:28]: Updated

File: EIPS/eip-5982.md [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+status: Draft
+type: Standards Track
+category: ERC
+created: 2022-11-15
+requires: 165, 5750
+---
+
+## Abstract
+
+This EIP defines an interface for role-based access control for smart contracts. Roles are defined as `byte32`. The interface specifies how to read, grant, create and destroy roles. It specifies the sense of role power in the format of its ability to call a given method
```
Comment (Pandapip1) [inline EIPS/eip-5982.md:16]:
```suggestion
This EIP defines an interface for role-based access control for smart contracts. Roles are the `keccak256` hash of the human-readable role name (in SCREAMING_SNAKE_CASE). The interface specifies how to read, grant, create and destroy roles.
```
I don't understand the last sentence. Mind explaining it?

Comment (xinbenlv) [reply EIPS/eip-5982.md:16]: > Roles are the `keccak256` hash of the human-readable role name (in SCREAMING_SNAKE_CASE).

While it's recommend to use `keccak256` upon a human readable string, this EIP _doesn't intend to mandate_ the choice of hash function. So I would probably not include this suggested update.

>> It specifies the sense of role power in the format of its ability to call a given method
>  I don't understand the last sentence. Mind explaining it?

I will revise that last sentence, it means this EIP not only define a way to get and set roles, but also defines a way to get the "power" of a role. The way to check if a role has certain power is by calling `canExecute` function and pass in a `bytes4` method selector in its second parameter

```
function canExecute(bytes32 executor, bytes4 method, bytes32 calldata payload, bytes calldata _data) view external returns(bool);
```

**Comment by xinbenlv** [comment]
WIP

**Comment by xinbenlv** [comment]
@frangio thank you for the feedback most of the functions are pretty self-explanatory and I was hoping to provide a condensed and more readable version for the first draft. I could add them before turning into `Review` status, if you don't mind?

**Comment by frangio** [comment]
Sure.

**Comment by xinbenlv** [comment]
@Pandapip1 this PR is ready to merge to my understanding~

**Review by Pandapip1 — APPROVED** [review]

**Pull request merged**